File: test/tile_coalesce_test.dart
import 'package:bloo_dot_evolutions/algo/monomino_index.dart';
import 'package:bloo_dot_evolutions/algo/monomino_lookup.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test("When_IterateAllNeighborConfigurations_Expect_CorrectHistogram", () async {
    Map<int, int> histogram = {};
    for (var nc = 0; nc <= 0xff; ++nc) {
      var tileIndex = MonominoIndex.coalescedTileFromNeighborConfiguration(nc);
      if (histogram.containsKey(tileIndex)) {
        histogram[tileIndex] = histogram[tileIndex]! + 1;
      } else {
        histogram[tileIndex] = 1;
      }
    }

    var distinctTiles = histogram.length;
    var distinctFrequencies = histogram.values.toSet().toList();
    distinctFrequencies.sort();

    expect(distinctTiles, MonominoLookup.clumsyPack.expand((element) => element).toSet().length);
    expect(distinctFrequencies, [1, 2, 4, 6, 8, 10, 12, 16, 20]);
  });
}

File: app/src/test/java/oy/sarjakuvat/flamingin/bde/algo/TileCoalesceTest.kt
package oy.sarjakuvat.flamingin.bde.algo

import org.junit.Assert.*
import org.junit.Test

class TileCoalesceTest {
    /* weird prerequisite for array literals */
    operator fun get(vararg array: Int) = array

    @Test
    fun when_iterateAllNeighborConfigurations_expect_correctHistogram() {
        val histogram = emptyMap<Int, Int>().toMutableMap()
        for (nc in 0..0xff) {
            val tileIndex = MonominoIndex.coalescedTileFromNeighborConfiguration(nc)
            if (histogram.containsKey(tileIndex)) {
                histogram[tileIndex] = histogram[tileIndex]!! + 1
            } else {
                histogram[tileIndex] = 1
            }
        }

        val distinctTiles = histogram.size
        val distinctFrequencies = histogram.values.toSet().toList()
        val sortedFrequencies = distinctFrequencies.sorted()
        var countInClumsy = 0
        for(row in 0 until MonominoLookup.clumsyPack.size) {
            countInClumsy += MonominoLookup.clumsyPack[row].size
        }

        assertEquals((this [1, 2, 4, 6, 8, 10, 12, 16, 20]).toList(), sortedFrequencies)
        assertEquals(countInClumsy, distinctTiles)
    }
}
